File: 51. N-Queens-1.cpp
class Solution {
public:
  vector<string> board;
  vector<int> col, diag1, diag2;
  vector<vector<string>> res;
  vector<vector<string>> solveNQueens(int n) {
    board = vector<string>(n, string(n, '.'));
    col = vector<int>(n, 0);
    diag1 = vector<int>(2 * n - 1, 0);
    diag2 = vector<int>(2 * n - 1, 0);

    helper(n, 0);

    return res;
  }

  void helper(int n, int i) {
    if (i == n) {
      res.push_back(board);
      return;
    }

    for (int j = 0; j < n; ++j) {
      if (!isValid(n, i, j))
        continue;
      update(n, i, j, true);
      helper(n, i + 1);
      update(n, i, j, false);
    }
  }

  void update(int n, int i, int j, bool isPut) {
    col[j] = isPut;
    diag1[i + j] = isPut;
    diag2[i - j + n - 1] = isPut;
    board[i][j] = isPut ? 'Q' : '.';
  }

  bool isValid(int n, int i, int j) {
    return !col[j] && !diag1[i + j] && !diag2[i - j + n - 1];
  }
};

File: 51. N-Queens-1.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []
        place = [0] * n
        col = [False] * n
        m = 2 * n - 1
        diag1 = [False] * m
        diag2 = [False] * m

        def dfs(i):
            if i == n:
                res.append(["." * j + "Q" + "." * (n - 1 - j) for j in place])
                return
            for j in range(n):
                if not col[j] and not diag1[i + j] and not diag2[i - j]:
                    place[i] = j
                    col[j] = diag1[i + j] = diag2[i - j] = True
                    dfs(i + 1)
                    col[j] = diag1[i + j] = diag2[i - j] = False

        dfs(0)
        return res
